webhook/aws/sns_handler.go

```diff
@@ -0,0 +1,243 @@
+  "SigningCertURL" : "https://sns.us-west-2.amazonaws.com/SimpleNotificationService-f3ecfb7224c7233fe7bb5f59f96de52f.pem",
+  "UnsubscribeURL" : "https://sns.us-west-2.amazonaws.com/?Action=Unsubscribe&SubscriptionArn=arn:aws:sns:us-west-2:123456789012:MyTopic:c9135db0-26c4-47ec-8998-413945fb5a96"
+  }
+*/
+type MsgAttr struct {
+	Type  string
+	Value string
+}
+type SNSMessage struct {
+	Type             string
+	MessageId        string
+	Token            string
+	TopicArn         string
+	Message          string
+	SubscribeURL     string
+	Timestamp        string
+	SignatureVersion string
+	Signature        string
+	SigningCertURL   string
+	UnsubscribeURL   string
+	MessageAttributes map[string]MsgAttr
+}
+
+// Define handlers for various AWS SNS POST calls
+func (ss *SNSServer) SetSNSRoutes(urlPath string, r *mux.Router, handler http.Handler) {
+	
+	r.HandleFunc(urlPath, ss.SubscribeConfirmHandle).Methods("POST").Headers("x-amz-sns-message-type", "SubscriptionConfirmation")
+	r.Handle(urlPath, handler).Methods("POST").Headers("x-amz-sns-message-type", "Notification")
+	r.HandleFunc(urlPath, ss.UnsubscribeConfirmHandle).Methods("POST").Headers("x-amz-sns-message-type", "UnsubscribeConfirmation")
+
+}
+
+// Subscribe to AWS SNS Topic to receive notifications
+func (ss *SNSServer) Subscribe() bool {
+	log := ss.logger()
+
+	params := &sns.SubscribeInput{
+		Protocol: aws.String(ss.SelfUrl.Scheme), // Required
+		TopicArn: aws.String(ss.Config.Sns.TopicArn), // Required
+		Endpoint: aws.String(ss.SelfUrl.String()),
+	}
+	log.Debug("SNS subscribe params: %#v", params)
+	resp, err := ss.SVC.Subscribe(params)
+	if err != nil {
+		log.Error("SNS subscribe error: %v", err)
+		return false
+	}
+
+	log.Debug("SNS subscribe resp: %v", resp)
+	
+	// Thread synchronization acquire lock and update
+	ss.UpdateSubscriptionArn(resp.SubscriptionArn)
+	
+	return true
+}
+
+// POST handler to receive SNS Confirmation Message
+func (ss *SNSServer) SubscribeConfirmHandle(rw http.ResponseWriter, req *http.Request) {
+	log := ss.logger()
+	defer func() {
+		if r := recover(); r != nil {
+			log.Error( "SNS SubscribeConfirmHandle recover error %v", r )
+		}
+	}()
+	msg := new(SNSMessage)
+
+	raw, err := DecodeJsonPayload(req, msg)
+	if err != nil {
+		log.Error("SNS read req body error %v", err)
+		return
+	}
+	//health.SendEvent(HTH.Set("TotalDataPayloadReceived", int(len(raw)) ))
+
+	log.Debug("SNS confirmation payload raw [%v]", string(raw))
+	
+	// TODO: Verify SNS Message authenticity by verifying signature
+
+	params := &sns.ConfirmSubscriptionInput{
+		Token:    aws.String(msg.Token),    // Required
+		TopicArn: aws.String(msg.TopicArn), // Required
+	}
+	resp, err := ss.SVC.ConfirmSubscription(params)
+	if err != nil {
+		log.Error("SNS confirm error %v", err)
+		// TODO return error response
+		return
+	}
+
+	log.Debug("SNS confirm response: %v", resp)
```

Comment:
I would also recommend logging things to be done in Decorators.  Log statements make the code noisy, but of course you still want to log.  Separating the logging into a Decorator makes this handler's code easier to read and lets you change what's logged without affecting your core business logic.